skpro/distributions/qpd.py

```diff
@@ -263,138 +265,78 @@ def __init__(
-        var = var_func(x, mean, cdf.T, index.shape[0])
-        return var
+        return ppf_arr
 
     def _pdf(self, x: np.ndarray):
-        """Probability density function.
+        """Probability density function."""
+        lower = self._bc_params["lower"]
+        delta = self._qpd_params["delta"]
+        kappa = self._qpd_params["kappa"]
+        c = self._qpd_params["c"]
+        n = self._qpd_params["n"]
+        theta = self._qpd_params["theta"]
 
-        this fucntion transform cdf to pdf
-        because j-qpd's pdf calculation is bit complex
-        """
-        return pdf_func(x, self.qpd)
+        phi = self.phi
 
-    def _ppf(self, p: np.ndarray):
-        """Quantile function = percent point function = inverse cdf."""
-        params = self.get_params(deep=False)
-        index = params["index"]
-        columns = params["columns"]
-        qv_low = params["qv_low"]
-        p_unique = np.unique(p)  # de-broadcast
-        ppf_all = ppf_func(p_unique, self.qpd)
-        ppf_map = np.tile(p_unique, (qv_low.size, 1)).T
-        ppf = np.zeros((index.shape[0], len(columns)))
-        for r in range(p.shape[0]):
-            for c in range(p.shape[1]):
-                t = np.where(ppf_map[:, c] == p[r][c])
-                ppf_part = ppf_all[t][c]
-                ppf[r][c] = ppf_part
-        return ppf
+        # we work through the chain rule for the entire nested expression in cdf
+        x_ = (x - lower) / theta
+        x_der = 1 / theta
+
+        in_arcsinh = np.log(x_) / kappa
+        in_arcsinh_der = x_der / (kappa * x_)
+
+        in_sinh = np.arcsinh(in_arcsinh) - np.arcsinh(n * c * delta)
+        in_sinh_der = arcsinh_der(in_arcsinh) * in_arcsinh_der
+
+        in_cdf = np.sinh(in_sinh) / delta
+        in_cdf_der = np.cosh(in_sinh) * in_sinh_der / delta
+
+        # cdf_arr = phi.cdf(in_cdf)
+        cdf_arr_der = phi.pdf(in_cdf) * in_cdf_der
+
+        pdf_arr = cdf_arr_der
+        return pdf_arr
 
     def _cdf(self, x: np.ndarray):
         """Cumulative distribution function."""
-        params = self.get_params(deep=False)
-        index = params["index"]
-        columns = params["columns"]
-        qv_low = params["qv_low"]
-        x_unique = np.unique(x)  # de-broadcast
-        cdf_all = cdf_func(x_unique, self.qpd)
-        cdf_map = np.tile(x_unique, (qv_low.size, 1)).T
-        cdf = np.zeros((index.shape[0], len(columns)))
-        for r in range(x.shape[0]):
-            for c in range(x.shape[1]):
-                t = np.where(cdf_map[:, c] == x[r][c])
-                cdf_part = cdf_all[t][c]
-                cdf[r][c] = cdf_part
-        return cdf
+        lower = self._bc_params["lower"]
+        delta = self._qpd_params["delta"]
+        kappa = self._qpd_params["kappa"]
+        c = self._qpd_params["c"]
+        n = self._qpd_params["n"]
+        theta = self._qpd_params["theta"]
+
+        phi = self.phi
+
+        in_arcsinh = np.log((x - lower) / theta) / kappa
```

Comment:
In most of the cases, you probably want to do a np.outer rather than an element-wise multiplication here. (In the Cyclic Boosting implementation I made that optional.)